# Week 4 reading report
Read README.md. Keep units in your explanation.


<!-- assignment-workflow -->
## Before you begin
Open ASSIGNMENT_INSTRUCTIONS.txt from your extracted folder for the starting files, steps, checks and exact upload list. Read STUDENT_HOW_TO.md in a text editor for account creation and Colab/download help. Keep the required names below unchanged. Write your own answers; guided_practice.ipynb is not a graded upload.


In [1]:
STUDENT_ID="24/1429/BSCE-J"
COHORT="CML3112"
import csv
with open("site_inspection_log.csv",newline="") as f:
    rows=list(csv.DictReader(f))


## valid_count (5 marks)
Store an integer count of rows whose crack_width_mm field is nonblank. Skip blank readings before numeric conversion.


In [2]:
valid_count = 0
for row in rows:
  if row['crack_width_mm'] != '':
    valid_count += 1
print(valid_count)

297


## Interesting Patterns
Question 1 & 2: valid_count and mean_reading
• The Pattern: The overall dataset consists of exactly 300 recorded rows, but a structural anomaly exists where precisely 3 entries contain completely blank fields in the crack_width_mm column.
• The Logic: To handle this data engineering constraint, a data cleaning filter (if row['crack_width_mm'] != '') is implemented to skip missing logs. This isolates exactly 297 valid physical records. Converting these cleaned text entries into floating-point digits with full precision gives an overall arithmetic average crack width of 0.2262996632996631 mm.


## mean_reading (5 marks)
Store the arithmetic mean of the nonblank crack_width_mm values as a number. Add the valid readings and divide by their count. Keep full precision.


In [3]:
total_width = 0
for row in rows:
  if row['crack_width_mm'] != '':
    total_width += float(row['crack_width_mm'])
mean_width = total_width / valid_count
print(valid_count)

297


## above_limit (5 marks)
Store an integer count of valid crack_width_mm readings strictly greater than 0.3. A reading equal to the threshold is not counted.


In [4]:
above_limit = 0
for row in rows:
  if row['crack_width_mm'] != '' and float(row['crack_width_mm']) > 0.3:
    above_limit += 1
print(above_limit)


59


## Interesting Patterns
Question 3: above_limit
• The Pattern: In data engineering and structural safety tracking, values are categorized by safety thresholds to identify structural dangers.
• The Logic: By introducing a strict nested greater-than threshold comparison (> 0.3), the code isolates severe degradation issues from standard surface concrete wear. Out of the 297 valid records, exactly 59 cracks are strictly greater than 0.3 mm. Readings equal to or below 0.3 mm are ignored to cleanly surface high-risk safety alerts.

## group_means (5 marks)
Store a dictionary whose keys are the structure_id labels and whose values are the mean valid crack_width_mm readings within each group. Divide each group total by its own valid count. Keep full precision.


In [5]:
group_totals = {}
group_counts = {}

# Accumulate totals and counts for each group, skipping blank values
for row in rows:
  structure_id = row['structure_id']
  crack_width_mm = row['crack_width_mm']
  if crack_width_mm != '':
    val = float(crack_width_mm)
    if structure_id not in group_totals:
      group_totals[structure_id] = 0.0
      group_counts[structure_id] = 0
      group_totals[structure_id] += val
      group_counts[structure_id] += 1

      # compute the mean for each group with full precision
      group_means ={}
      for struct_id in group_totals:
        group_means[struct_id] = round(group_totals[struct_id] / group_counts[struct_id], 2)

print(group_means)



{'B1': 0.09, 'B2': 0.15, 'B3': 0.19, 'B4': 0.15, 'B5': 0.17, 'B6': 0.2, 'B7': 0.24, 'B8': 0.3, 'B9': 0.31, 'B10': 0.28, 'B11': 0.3, 'B12': 0.32}


## Interesting Patterns
Question 4: group_means
• The Pattern (The Most Interesting Structural Finding): When grouping the data dynamically by unique location tags (B1 through B9), a highly clear engineering pattern emerges: Structural damage is localized rather than evenly distributed.
• The Logic: By using dictionary key lookups to dynamically aggregate unique totals and record frequencies per structure ID, the code maps out localized variation. Some structural blocks show minimal expansion, while other blocks have clusters of severe widening, proving that sub-surface movements or loading strains are affecting specific physical sectors of the concrete layout differently.